src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+    this.records.push(rec);
+    return rec.proxy();
+  }
+
+  record = (...args:any[]) => {
+    let out = this.find.apply(null, args);
+    out.output = true;
+    return out;
+  }
+
+  registerVariable(variable:DSLVariable) {
+    this.variableLookup[variable.id] = [variable];
+  }
+
+  toValue(a:any) {
+    if(a === undefined || a === null) throw new Error("Eve values can't be undefined or null");
+    if(a instanceof DSLVariable) {
+      return a.value;
+    } if(a instanceof DSLRecord) {
+      return a.record.value;
+    } if(a instanceof DSLFunction) {
+      return a.returnValue.value;
+    }
+    return a;
+  }
+
+  equivalence(a:any, b:any) {
+    let aValue = a.value || (a.record && a.record.value) || a;
+    let bValue = b.value || (b.record && b.record.value) || b;
+    let aIsRegister = isRegister(aValue);
+    let bIsRegister = isRegister(bValue);
+    if(aIsRegister && bIsRegister) {
+      let aVars = this.variableLookup[a.id];
+      let bVars = this.variableLookup[b.id];
+      for(let variable of aVars) {
+        variable.value = bValue;
+        bVars.push(variable);
+      }
+      this.variableLookup[a.id] = [];
+    } else if(aIsRegister) {
+      let aVars = this.variableLookup[a.id];
+      for(let variable of aVars) {
+        variable.value = bValue;
+      }
+      this.variableLookup[a.id] = [];
+    } else if(bIsRegister) {
+      let bVars = this.variableLookup[b.id];
+      for(let variable of bVars) {
+        variable.value = aValue;
+      }
+      this.variableLookup[b.id] = [];
+    } else if(aValue !== bValue) {
+      throw new Error(`Trying to equivalence two static values that aren't the same: ${aValue} and ${bValue}`);
+    }
+  }
+
+  prepare() {
+    let functions = this.functions.slice() as (DSLFunction | undefined)[];
+    // look through the functions for equivalences before we do anything else.
+    let ix = 0;
+    for(let func of functions) {
+      if(!func || func.path[1] !== "==") continue;
+      this.equivalence(func.args[0], func.args[1]);
+      functions[ix] = undefined;
+      ix++;
+    }
+    let registerIx = 0;
+    for(let id in this.variableLookup) {
+      let variable = this.variableLookup[id][0] as DSLVariable;
+      if(!variable || !isRegister(variable.value)) continue;
+      if(variable.value.offset !== -1) throw new Error("We've somehow already assigned a variable's register");
+      variable.value.offset = registerIx++;
+    }
+    let items = functions.concat(this.records as any[]);
+    let constraints = [];
+    let nodes = [];
+    for(let toCompile of items) {
+      if(!toCompile) continue;
+      let compiled = toCompile.compile();
+      if(!compiled) continue;
+      for(let item of compiled) {
+        if(item instanceof Scan || item instanceof FunctionConstraint) {
+          constraints.push(item);
+        } else {
+          nodes.push(item as Node);
+        }
+      }
+    }
+    nodes.unshift(new runtime.JoinNode(constraints))
```

Comment:
I know we don't have aggregates yet, but I assume once we do we'll need to be a little smarter here and stratify across those boundaries into separate JoinNodes?